### PR TITLE
Vertical transformation of features in iterators

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR adds API to ensure that z values in features are correctly transformed by iterators when a coordinate transform involving vertical components is in effect.

### CODE REVIEW

File: tests/src/python/test_qgsfeatureiterator.py [hunk 1]
```diff
@@ -533,6 +537,196 @@ def callback(feature):
         self.assertEqual(res, ['a', 'b'])
         layer.rollBack()
 
+    def test_vertical_transformation_gda2020_to_AVWS(self):
+        """
+        Test vertical transformations are correctly handled during iteration
```
Comment (rouault) [inline tests/src/python/test_qgsfeatureiterator.py:542]: Those tests are fragile as they require grids to be present, and may depend on the actual PROJ version. Perhaps just testing a transformation between WGS 84 geocentric (EPSG:4978) and WGS 84 3D (EPSG:4979) would be more robust.
Or passing a dummy PROJ pipeline for example with a +proj=affine +xoff=... +yoff=.... +zoff=.... (https://proj.org/en/9.4/operations/transformations/affine.html)
It is also not obvious to me those 4 slightly different tests test different aspect of this PR (some of them might test aspects already implemented regarding 3D / compound CRS)

Comment (nyalldawson) [reply tests/src/python/test_qgsfeatureiterator.py:542]: @rouault what if I made them skipped if the grid isn't available? 

I **really** want tests which check real world conversions which match with officially published results.

Comment (nyalldawson) [reply tests/src/python/test_qgsfeatureiterator.py:542]: To answer the second part of your question -- they are "katamari" tests which check that the full set of vertical crs handling is working. They'll fail without any of the commits in this pr, but also depend on earlier commits.

Comment (rouault) [reply tests/src/python/test_qgsfeatureiterator.py:542]: your call... but that's still fragile. For example if a new geoid model is released in the future for EPSG:5711 / AHD height, the results could change.

Comment (nyalldawson) [reply tests/src/python/test_qgsfeatureiterator.py:542]: Ok. I'm leaning towards adding the stable test using 4978/4979, and then leaving the existing ones for security. 

(I once worked in an office where someone screwed up a multi-million dollar project by messing up a vertical transform ... So I'm kinda paranoid here 😱)

**Comment by nyalldawson** [comment]
Hmm, the crashing vector layer cache test isn't caused by this PR . I've fixed the underlying issue in https://github.com/qgis/QGIS/pull/58093

**Review by rouault — APPROVED** [review]

**Pull request merged**